### PR TITLE
Port a few rules from Diffractor

### PR DESCRIPTION
**Pull request by Keno**

### CODE REVIEW

File: src/rulesets/Core/core.jl
Comment (simeonschaub) [inline src/rulesets/Core/core.jl:7]:
```suggestion
@non_differentiable Core.apply_type(::Any, ::Any...)
```

File: src/rulesets/Core/core.jl
Comment (simeonschaub) [inline src/rulesets/Core/core.jl:12]:
```suggestion
    typeassert_pullback(Δ) = (NoTangent(), Δ, NoTangent())
    return typeassert(x, T), typeassert_pullback
```

File: src/rulesets/Core/core.jl
Comment (simeonschaub) [inline src/rulesets/Core/core.jl:17]:
```suggestion
    ifelse_pullback(Δ) = (NoTangent(), NoTangent(), ifelse(c, Δ, ZeroTangent()), ifelse(c, ZeroTangent(), Δ))
    return ifelse(c, a, b), ifelse_pullback
```

File: src/rulesets/Core/core.jl
Comment (simeonschaub) [inline src/rulesets/Core/core.jl:18]:
```suggestion
end

```

File: test/rulesets/Core/core.jl
Comment (simeonschaub) [inline test/rulesets/Core/core.jl:9]:
```suggestion
end

```

**Review by simeonschaub — APPROVED** [review]
Some small style nits, but otherwise LGTM.

File: src/rulesets/Core/core.jl
Comment (simeonschaub) [reply src/rulesets/Core/core.jl:7]: Or maybe even
```suggestion
@non_differentiable Core.apply_type(::Type, ::Any...)
```
?

File: src/rulesets/LinearAlgebra/structured.jl
Comment (mzgubic) [inline src/rulesets/LinearAlgebra/structured.jl:101]: Just an FYI: we are considering handling this at the AD level, see: https://github.com/JuliaDiff/ChainRules.jl/issues/442#issuecomment-872098608

But that is currently blocked by [`ProjectTo` PR](https://github.com/JuliaDiff/ChainRulesCore.jl/pull/385) which I am working on at the moment.

File: src/rulesets/Base/array.jl
Comment (mzgubic) [inline src/rulesets/Base/array.jl:16]: Does this not work? https://github.com/JuliaDiff/ChainRulesCore.jl/blob/653cd7d626183faf70107a6c639c04ae4a85f0c5/src/differentials/thunks.jl#L49

File: src/rulesets/Core/core.jl
Comment (mcabbott) [reply src/rulesets/Core/core.jl:17]: It's possible there should be another method for `rrule(::typeof(ifelse), c, a::Number, b::Number)` with `ifelse_pullback_2(Δ) = (NoTangent(), NoTangent(), ifelse(c, Δ, zero(Δ)), ifelse(c, zero(Δ), Δ))`, to be type-stable.

File: src/rulesets/Base/array.jl
Comment (simeonschaub) [reply src/rulesets/Base/array.jl:16]:
```suggestion
        ∂A = reshape(Ȳ, A_dims)
```

File: src/rulesets/LinearAlgebra/structured.jl
Comment (mcabbott) [reply src/rulesets/LinearAlgebra/structured.jl]: Should these lines be removed before this is merged? Or commented `# remove me once X`?

Comment (mzgubic) [reply src/rulesets/LinearAlgebra/structured.jl]: Ideally removed, but it needs Diffractor to automatically propagate AbstractZeros, something like Zygote's https://github.com/FluxML/Zygote.jl/blob/5b803d12f2d235755ba6f23a9017fc29fefd9985/src/compiler/chainrules.jl#L143

If Diffractor does not do that, we should make an issue about it (and link it here)

Comment (simeonschaub) [reply src/rulesets/LinearAlgebra/structured.jl]: So is it fine to leave this in or should I just remove that part for now?

File: src/rulesets/Core/core.jl
Comment (simeonschaub) [reply src/rulesets/Core/core.jl]: Ah, wait, this can also be `TypeofVararg`!

File: src/rulesets/Core/core.jl
Comment (simeonschaub) [inline src/rulesets/Core/core.jl:10]:
```suggestion
@non_differentiable Core.apply_type(::Any, ::Any...)
```

File: src/rulesets/LinearAlgebra/structured.jl
Comment (mzgubic) [reply src/rulesets/LinearAlgebra/structured.jl]: Yeah let's make sure it all works for Diffractor and just link this: https://github.com/JuliaDiff/Diffractor.jl/issues/25

File: test/rulesets/Core/core.jl
Comment (mzgubic) [inline test/rulesets/Core/core.jl:11]:
```suggestion
    test_rrule(ifelse, true, 1.1, 2.0)
    test_frule(ifelse, false, 1.1, 2.0)

    test_rrule(ifelse, true, [1.1], [2.0]; check_inferred=false)
    test_frule(ifelse, false, [1.1], [2.0]; check_inferred=false)
```
`true`/`false` automatically gets a `NoTangent()`. And `1.0`s and `0.0`s are sometimes tricky inputs as they spuriously give the right answer

File: test/rulesets/Core/core.jl
Comment (mzgubic) [inline test/rulesets/Core/core.jl:3]:
```suggestion
    test_rrule(typeassert, 1.1, Float64)
    test_frule(typeassert, 1.1, Float64)
```

File: src/rulesets/Base/nondiff.jl
Comment (mzgubic) [inline src/rulesets/Base/nondiff.jl:471]:
```suggestion
@non_differentiable eltype(::Type)
```

could we also move the eltype to the right place alphabetically?

**Review by mzgubic — APPROVED** [review]
Thanks, looks good! Feel free to merge when the style comments are addressed

File: src/rulesets/Base/nondiff.jl
Comment (simeonschaub) [reply src/rulesets/Base/nondiff.jl:471]: Do you mean
```suggestion
@non_differentiable eltype(::Type)
@non_differentiable similar(::Any...)

@non_differentiable one(::Any)
@non_differentiable ones(::Any...)
@non_differentiable zero(::Any)
@non_differentiable zeros(::Any...)
```
?

Comment (mzgubic) [reply src/rulesets/Base/nondiff.jl]: No, sorry, I did mean to delete the others, since they are already in base.jl on master, see e.g. https://github.com/JuliaDiff/ChainRules.jl/blob/3f18cd6944378f16827b0e47bc3aa4a264caed55/src/rulesets/Base/nondiff.jl#L307
```julia
julia> @which ones
Base
```

Comment (mzgubic) [reply src/rulesets/Base/nondiff.jl]: actually eltype also seems to belong to Base

Comment (simeonschaub) [reply src/rulesets/Base/nondiff.jl]: Ah, sorry,  I misunderstood. `eltype` should still be in `Base/nondiff.jl` though, right?

Comment (mzgubic) [reply src/rulesets/Base/nondiff.jl]: Correct! My bad, forgot that `Base/base.jl` exists 😄

File: Project.toml
Comment (mzgubic) [inline Project.toml:3]: ah, sorry, just saw this - I think we want to increment the minor version for adding rules, and the patch if we are fixing bugs in existing rules.

```suggestion
version = "1.3.0"
```

**Review by mzgubic — APPROVED** [review]

**Pull request merged**
